app/controllers/automated_tests_controller.rb

```diff
@@ -35,12 +35,26 @@ def student_interface
     @student = current_role
     @grouping = @student.accepted_grouping_for(@assignment.id)
 
-    unless @grouping.nil?
+    if @grouping.nil?
+      @next_token_generation_time = nil
+    else
       @grouping.refresh_test_tokens
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
+
+      if @assignment.enable_student_tests
+        # Calculate the next token generation time
+        token_period = @assignment.token_period
+        last_generated = @grouping.test_runs.first&.created_at || Time.zone.now
+        @next_token_generation_time = last_generated + token_period.hours
+
+        # Format the next token generation time for display
+        @next_token_generation_time = @next_token_generation_time.strftime('%A, %B %d, %Y %I:%M:%S %p %Z')
```

Comment:
You can use `I18n.l` in controllers to do the same type of date formatting as in views